File: c++/2295.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

int arr[1000];
vector<int> sum;

bool binarySearch(int target) {
  int start = 0;
  int end = sum.size() - 1;

  while (start <= end) {
    int mid = (start + end) / 2;

    if (target < sum[mid])
      end = mid - 1;
    else if (target > sum[mid])
      start = mid + 1;
    else
      return true;
  }

  return false;
}

int main() {
  freopen("example.txt", "r", stdin);
  int N;
  cin >> N;

  for (int i = 0; i < N; i++) {
    cin >> arr[i];
  }

  for (int i = 0; i < N; i++) {
    for (int j = i; j < N; j++) {
      sum.push_back(arr[i] + arr[j]);
    }
  }

  sort(arr, arr + N);
  sort(sum.begin(), sum.end());

  for (int k = N - 1; k >= 0; k--) {
    for (int z = k; z >= 0; z--) {
      int target = arr[k] - arr[z];
      if (binarySearch(target)) {
        cout << arr[k];
        return 0;
      }
    }
  }

  return 0;
}

File: js/2295.js
const fs = require('fs');

const inputs = fs.readFileSync('example.txt').toString().trim().split('\n').map(Number);

const [N, ...arr] = inputs;

const sum = [];

for (let i = 0; i < N; i++) {
  for (let j = i; j < N; j++) {
    sum.push(arr[i] + arr[j]);
  }
}

arr.sort((a, b) => a - b);
sum.sort((a, b) => a - b);

c: for (let k = N - 1; k >= 0; k--) {
  for (let z = k; z >= 0; z--) {
    const target = arr[k] - arr[z];

    if (binarySearch(target)) {
      console.log(arr[k]);
      break c;
    }
  }
}

function binarySearch(target) {
  let start = 0;
  let end = sum.length - 1;

  while (start <= end) {
    const mid = Math.floor((start + end) / 2);

    if (sum[mid] > target) end = mid - 1;
    else if (sum[mid] < target) start = mid + 1;
    else return true;
  }

  return false;
}
